### ISSUE TITLE
Calendar export fails with Sabre "End of document reached prematurely" Exception

### ISSUE THREAD

**Issue by ervee** [owncloud/calendar#861]
### Steps to reproduce
1. Click the download calendar arrow next to my default calendar.

### Expected behaviour
I should be presented with an ICS file to download. Other calendar id's download fine, just not this calendar id due to some Sabre error.

This calendar does show fine in the OC calendar!
It also syncs with the CardDAV-Sync free App from Marten Gajda!
It does NOT sync with the DAVdroid App from bitfire web engineering . With this App I get a huge amount if these errors in the Apache log: [Fri Jul 17 15:24:21 2015] [error] [client remote_ip] {"reqId":"some_id","remoteAddr":"remote_ip","app":"calendar","message":"Event (some_ID_string-caldavsyncadapter.ics) contains invalid data: End of document reached prematurely","level":2,"time":"2015-07-17T15:24:21+02:00"}, referer: https://my_domain/cloud/index.php/apps/calendar/

### Actual behaviour
Firefox shows:

File not found

Firefox can't find the file at https://my_domain/cloud/index.php/apps/calendar/export.php?calid=1.

Check the file name for capitalization or other typing errors.
Check to see if the file was moved, renamed or deleted.

### Server configuration
Debian GNU/Linux 7 (wheezy)

**Web server:**
Apache2 version: 2.2.22-13+deb7u4

**Database:**
MariaDB version: 5.5.44+maria-1~wheezy

**PHP version:**
5.4.41-0+deb7u1

**ownCloud version:** (see ownCloud admin page)
8.1.0 (stable)

**Updated from an older ownCloud or fresh install:**
Updated

**List of activated apps:**

```
 - calendar: 0.7.1
 - contacts: 0.4.0.0
 - files: 1.1.9
 - files_locking:
 - files_pdfviewer: 0.7
 - files_sharing: 0.6.2
 - files_texteditor: 0.4
 - files_trashbin: 0.6.3
 - files_versions: 1.0.6
 - files_videoviewer: 0.1.3
 - firstrunwizard: 1.1
 - galleryplus: 12
 - mozilla_sync: 1.4
 - search_lucene: 0.6.0
 - user_pwauth: 2.1
 - youtubedl: 0.0.4
```

**The content of config/config.php:**

```
<?php
$CONFIG = array (
  'instanceid' => '529ce42a91e36',
  'passwordsalt' => '**********',
  'datadirectory' => '/media/Cloud',
  'dbtype' => 'mysql',
  'version' => '8.1.0.8',
  'dbname' => 'cloud',
  'dbhost' => 'my_db_host.my_domain',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'oc_admin',
  'dbpassword' => '**********',
  'installed' => true,
  'theme' => '',
  'maintenance' => false,
  'forcessl' => true,
  'logfile' => '/var/log/owncloud.log',
  'loglevel' => '1',
  'logtimezone' => 'Europe/Amsterdam',
  'log_authfailip' => true,
  'mail_domain' => 'my_domain',
  'mail_smtpmode' => 'smtp',
  'mail_smtphost' => 'smtp.my_domain',
  'mail_smtpport' => '25',
  'mail_smtptimeout' => '10',
  'trusted_domains' =>
  array (
    0 => 'my_domain',
  ),
  'share_folder' => '/Shared',
  'overwritewebroot' => '/cloud',
  'overwrite.cli.url' => '/cloud',
  'mail_from_address' => 'webmaster',
  'enable_previews' => true,
  'preview_max_x' => NULL,
  'preview_max_y' => NULL,
  'preview_max_scale_factor' => 1,
  'enabledPreviewProviders' =>
  array (
    0 => 'OC\\Preview\\Image',
    1 => 'OC\\Preview\\Illustrator',
    2 => 'OC\\Preview\\Postscript',
    3 => 'OC\\Preview\\Photoshop',
    4 => 'OC\\Preview\\TIFF',
  ),
  'preview_libreoffice_path' => '/usr/bin/libreoffice',
  'secret' => '**********',
  'asset-pipeline.enabled' => true,
);

```

**Are you using external storage, if yes which one:** local/smb/sftp/...
Storage is on an NFS mount.

**Are you using encryption:** yes/no
no

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
user_pwauth (local Unix auth)

### Client configuration
**Browser:**
Firefox 39.0

**Operating system:**
Windows 7 x64

#### Web server error log
```
Access log:
[17/Jul/2015:15:25:14 +0200] 212.61.144.165 TLSv1.2 ECDHE-RSA-AES128-GCM-SHA256 "GET /cloud/index.php/apps/calendar/export.php?calid=1 HTTP/1.1" 2497 "https://my_domain/cloud/index.php/apps/calendar/" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:39.0) Gecko/20100101 Firefox/39.0"

Error log:
[Fri Jul 17 15:25:15 2015] [error] [client 212.61.144.165] {"reqId":"\\/dCwcSzlLsuW6RAZ7S3C","remoteAddr":"212.61.144.165","app":"index","message":"Exception: {\\"Exception\\":\\"Sabre\\\\\\\\VObject\\\\\\\\EofException\\",\\"Message\\":\\"End of document reached prematurely\\",\\"Code\\":0,\\"Trace\\":\\"#0 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/3rdparty\\\\\\/sabre\\\\\\/vobject\\\\\\/lib\\\\\\/Parser\\\\\\/MimeDir.php(168): Sabre\\\\\\\\VObject\\\\\\\\Parser\\\\\\\\MimeDir->readLine()\\\\n#1 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/3rdparty\\\\\\/sabre\\\\\\/vobject\\\\\\/lib\\\\\\/Parser\\\\\\/MimeDir.php(137): Sabre\\\\\\\\VObject\\\\\\\\Parser\\\\\\\\MimeDir->parseLine('BEGIN:VEVENT')\\\\n#2 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/3rdparty\\\\\\/sabre\\\\\\/vobject\\\\\\/lib\\\\\\/Parser\\\\\\/MimeDir.php(63): Sabre\\\\\\\\VObject\\\\\\\\Parser\\\\\\\\MimeDir->parseDocument()\\\\n#3 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/3rdparty\\\\\\/sabre\\\\\\/vobject\\\\\\/lib\\\\\\/Reader.php(44): Sabre\\\\\\\\VObject\\\\\\\\Parser\\\\\\\\MimeDir->parse('BEGIN:VCALENDAR...', 0)\\\\n#4 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/apps\\\\\\/calendar\\\\\\/lib\\\\\\/export.php(68): Sabre\\\\\\\\VObject\\\\\\\\Reader::read('BEGIN:VCALENDAR...')\\\\n#5 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/apps\\\\\\/calendar\\\\\\/lib\\\\\\/export.php(43): OC_Calendar_Export::generateEvent(Array)\\\\n#6 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/apps\\\\\\/calendar\\\\\\/lib\\\\\\/export.php(28): OC_Calendar_Export::calendar('1')\\\\n#7 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/apps\\\\\\/calendar\\\\\\/export.php(20): OC_Calendar_Export::export('1', 'calendar')\\\\n#8 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/lib\\\\\\/private\\\\\\/route\\\\\\/route.php(154) : runtime-created function(1): require_once('\\\\\\/var\\\\\\/www\\\\\\/ownclo...')\\\\n#9 [internal function]: __lambda_func(Array)\\\\n#10 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/lib\\\\\\/private\\\\\\/route\\\\\\/router.php(274): call_user_func('\\\\\\\\x00lambda_50', Array)\\\\n#11 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/lib\\\\\\/base.php(867): OC\\\\\\\\Route\\\\\\\\Router->match('\\\\\\/apps\\\\\\/calendar\\\\\\/...')\\\\n#12 \\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/index.php(40): OC::handleRequest()\\\\n#13 {main}\\",\\"File\\":\\"\\\\\\/var\\\\\\/www\\\\\\/owncloud\\\\\\/3rdparty\\\\\\/sabre\\\\\\/vobject\\\\\\/lib\\\\\\/Parser\\\\\\/MimeDir.php\\",\\"Line\\":250}","level":4,"time":"2015-07-17T15:25:15+02:00"}, referer: https://my_domain/cloud/index.php/apps/calendar/
```

#### ownCloud log (data/owncloud.log)
```
Does not log a thing (set level to 0 (debug))
```

#### Browser log
```
File not found

Firefox can't find the file at https://my_domain/owncloud/index.php/apps/calendar/export.php?calid=1.

Check the file name for capitalization or other typing errors.
Check to see if the file was moved, renamed or deleted.
```

**Comment by georgehrke** [comment]
> End of document reached prematurely

Some event in your calendar is broken.
see https://github.com/owncloud/calendar/issues/680#issuecomment-73427922 on how to find the broken one.

**Comment by ervee** [comment]
When I run this command I get 814 results (only in calendar ID 1). What can I do to fix them? I can't just delete them all :)

**Comment by ervee** [comment]
My calendar ID 1 has a total of 958 events and 814 don't end with END:VCALENDAR ? I hope someone has a fix-script for this or I have to manualy edit 814 entries and put END:VCALENDAR at the end? :(

**Comment by ervee** [comment]
@georgehrke , do you think you can find anyone in the OwnCloud team who can whip up a script to fix entries not containing END:VCALENDAR ?

**Comment by georgehrke** [comment]
You can most certainly fix most of them by appending
```
END:VEVENT
END:VCALENDAR
```
(different for todos and journals).
That would "restore" the events, but most of them would miss information and be incomplete.

**Comment by ervee** [comment]
I exported all the erroneous events (with the SELECT * FROM oc_clndr_objects WHERE calendardata NOT LIKE '%END:VCALENDAR%'; query) and checked them. They all had no real event data in them and all 814 events were from the same creation time. So something went wrong there...

I executed this after testing in a test database:
DELETE FROM oc_clndr_objects WHERE calendardata NOT LIKE '%END:VCALENDAR%';

And now the erroneous events are gone and also the Sabre errors in Apache stopped. And the calendar export works again! Case closed.
